src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -0,0 +1,370 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics.Log;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Shared.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    internal partial class DiagnosticIncrementalAnalyzer
+    {
+        /// <summary>
+        /// This is responsible for getting diagnostics for given input.
+        /// It either return one from cache or calculate new one.
+        /// </summary>
+        private class Executor
+        {
+            private readonly DiagnosticIncrementalAnalyzer _owner;
+
+            public Executor(DiagnosticIncrementalAnalyzer owner)
+            {
+                _owner = owner;
+            }
+
+            public IEnumerable<DiagnosticData> ConvertToLocalDiagnostics(Document targetDocument, IEnumerable<Diagnostic> diagnostics, TextSpan? span = null)
+            {
+                var project = targetDocument.Project;
+
+                if (project.SupportsCompilation)
+                {
+                    return ConvertToLocalDiagnosticsWithCompilation(targetDocument, diagnostics, span);
+                }
+
+                return ConvertToLocalDiagnosticsWithoutCompilation(targetDocument, diagnostics, span);
+            }
+
+            public async Task<DocumentAnalysisData> GetDocumentAnalysisDataAsync(
+                CompilationWithAnalyzers analyzerDriverOpt, Document document, StateSet stateSet, AnalysisKind kind, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    var version = await GetDiagnosticVersionAsync(document.Project, cancellationToken).ConfigureAwait(false);
+                    var state = stateSet.GetActiveFileState(document.Id);
+                    var existingData = state.GetAnalysisData(kind);
+
+                    if (existingData.Version == version)
+                    {
+                        return existingData;
+                    }
+
+                    // perf optimization. check whether analyzer is suppressed and avoid getting diagnostics if suppressed.
+                    // REVIEW: IsAnalyzerSuppressed call seems can be quite expensive in certain condition. is there any other way to do this?
```

Comment:
the comment is based on looking at the implementation. it does linear search, and we have many place calling it without any cache. so if number of supposed diagnostics gets bigger (for each analyzer or total of analyzer), perf of all those places will also linearly decrease.

right now, we have one known such analyzer - compiler analyzer, and we special cases it, but any other user analyzer can cause same effect as compiler analyzer.